lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,90 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+void generateMemberwiseInit(ClassDecl *classDecl,
+                            SourceEditConsumer &EditConsumer,
+                            SourceManager &SM) {
+  
+  swift::SourceLoc targetLocation =
+  classDecl->getBraces().Start.getAdvancedLoc(1);
+  
+  SmallVector<std::string, 5> memberNameVector;
+  SmallVector<std::string, 5> memberTypeVector;
+  
+  for (auto varDecl : classDecl->getStoredProperties()) {
+    auto parentPatternBinding = varDecl->getParentPatternBinding();
+    if (!parentPatternBinding) continue;
+    
+    auto varDeclIndex =
+    parentPatternBinding->getPatternEntryIndexForVarDecl(varDecl);
+    
+    if (auto init = varDecl->getParentPatternBinding()->getInit(varDeclIndex)) {
+      if (init->getStartLoc().isValid())
+      continue;
+    }
+
+    StringRef memberName = varDecl->getName().str();
+    memberNameVector.push_back(memberName.str());
+    
+    std::string memberType = varDecl->getType().getString();
+    memberTypeVector.push_back(memberType);
+  }
+  
+  if (memberTypeVector.size() > 0) {
+    EditConsumer.accept(SM, targetLocation, "\ninternal init(");
+    
+    for (size_t i = 0; i < memberTypeVector.size(); i++) {
+      EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " + memberTypeVector[i]);
+      
+      if (i != memberTypeVector.size() - 1) {
+        EditConsumer.accept(SM, targetLocation, ", ");
+      }
+    }
+    
+    EditConsumer.accept(SM, targetLocation, ") {\n\n");
+    
+    for (auto varName: memberNameVector) {
+      EditConsumer.accept(SM, targetLocation,
+                          "self." + varName + " = " + varName + "\n");
+    }
+    
+    EditConsumer.accept(SM, targetLocation, "\n}\n");
+  }
+}
+  
+ClassDecl* canGenerateMemberwiseInit(ResolvedCursorInfo CursorInfo) {
+  
+  ClassDecl *classDecl = dyn_cast<ClassDecl>(CursorInfo.ValueD);
+  if (!classDecl || classDecl->getStoredProperties().empty()) {
```

Comment:
@nkcsgexi is there a way to determine this cursor is at actually the declaration?
For example, what if `typealias Alias = /* cursorinfo here->*/TargetClass` and `TargetClass` is declared in other file?

Reply:
`ResolvedCursorInfo.IsRef` indicates whether the cursor is on a reference or a declaration name. I think this refactoring should be only available when `ResolvedCursorInfo.IsRef` is false.